### DISCUSSION TITLE
Help integrating a vector function

### DISCUSSION THREAD

**Discussion by piperfw** [nschloe/quadpy#422] [Q&A] [answered]
Hi, I'm looking to integrate a vector-valued function of a single variable. Following the example
```python
def integrand(x):
    return [numpy.sin(x), numpy.cos(x)]

res, err = quadpy.quad(integrand, 0, 1)
```
I tried 
```python
def integrand(x):
    return [numpy.sin(x), 1.0]

res, err = quadpy.quad(integrand, 0, 1)
```
Expecting the second element of `res` to be `1.0`, but instead I get the ValueError
`ValueError: Expected the function return value to be of shape (..., 21), but got shape (2,) instead.`

Where am I going wrong? Apologies for such a basic question. I noted that changing `1.0` to `1.0+x*0.0` allowed the second integral to be performed.

**Comment by nschloe** [comment, marked as answer]
```python
import numpy as np
import quadpy


def integrand(x):
    return [np.sin(x), np.ones_like(x)]

res, err = quadpy.quad(integrand, 0, 1)
```

**Reply by piperfw** [reply]
I see, thank you for answering so quickly. Might it be worth having this example and a note on the input/output wiki page? I don't think it is obvious (or indeed from the error produced) that this is required.

**Reply by piperfw** [reply]
I am finding this quite restrictive still. For example, I don't seem to be able to integrate a vector function that involves matrix multiplication:
```python
def integrand(t):
    return np.matmul(2*np.array([[t*0.0,1.0*t],[t*1.0,t*0.0]]), np.array([[t*1.0,0.0*t],[0.0*t,1.0*t]]))
```
Is it possible to tell `quadpy` exactly what the input and output dimensions should be to avoid these issues? I've tried setting the optional parameters `domain_shape=(1,)`, `range_shape=(2,2)` (for the above integrand) but to no avail.

**Reply by nschloe** [reply]
Have you checked https://github.com/nschloe/quadpy/wiki/Dimensionality-of-input-and-output-arrays yet?

**Reply by piperfw** [reply]
Yes, I think in my problem `d=1`, `n=1` and `p` is determined behind the scenes (21?), but I am not sure how to use this information to avoid the errors.

**Reply by nschloe** [reply]
What's in shape of `t` in your case and what are you trying to compute?

**Reply by piperfw** [reply]
`t` is just a scalar, so `(1,)`. I'm simply trying to integrate element-wise a function that takes a scalar as an input (`t`), performs some operations on `2x2` matrices (with `t` dependence), and then returns the result (in fact in my full problem I have a kroncker product of 2x2 matrices, so the output is of shape `(4,4)` - I obviously want to avoid having to define 16 functions as the elements of this vector function when integrating).

**Reply by nschloe** [reply]
> t is just a scalar, so (1,)

That's not a scalar, that's a vector of length 1. Are you sure that this is the shape of `t`? That'd be extremely unusual. What integration scheme are you using? I will need to see an MWE.

**Reply by piperfw** [reply]
Thanks, I was just doing a MWE for you:
```python
#!/usr/bin/env python
import numpy as np
import quadpy

A=np.array([[0.0,1.0],[1.0,0.0]])
B=np.array([[1.0,0.0],[0.0,-1.0]])
def comm(mat):
    return np.kron(mat,np.eye(2)) - np.kron(np.eye(2),mat.T)
def f(t):
    return 2*t*A + B
def g(t):
    return comm(f(t))
g(1) # 4x4 matrix
quadpy.quad(g, 0, 1) # integrate each element component-wise
```
Ah, I misunderstood, `t` is a float so doesn't have a shape? I have no particular integration scheme in mind.

**Reply by nschloe** [reply]
With `import`s, please. I just want to copy and paste and note waste time guessing the modules.

**Reply by nschloe** [reply]
So, `t` is of shape `(21,)`, i.e., you're asked to return your function evaluated at 21 points. You can do that via a loop to start with. Later, if you're about speed, you want to vectorize it. This
```python
    out = np.array([comm(f(T)) for T in t])
```
has shape `(21, 4, 4)`, but you need shape `(..., 21)`. One solution is to just move the first axis to the end:
```
def g(t):
    out = np.array([comm(f(T)) for T in t])
    return np.moveaxis(out, 0, -1)
```
After vectorization, all these shenanigans will disappear and your code will be readable and fast.
Anyway, the result is
```
[[ 0. -1.  1.  0.]
 [-1.  2.  0.  1.]
 [ 1.  0. -2. -1.]
 [ 0.  1. -1.  0.]]
```

**Reply by piperfw** [reply]
@nschloe that is really helpful, thank you for working on this problem with me. May I check - the number '21' isn't magic but just some minimum number of points the integration scheme requires in order to ensure small error? So in more difficult cases one would expect it to be larger.

**Reply by nschloe** [reply]
21 Is the default number of points in the integration scheme; it's the 21-point Gauss–Kronrod quadrature. Because this integral here is so trivial, it doesn't need to subdivide the interval. If the function is more complex, expect many more points.

**Reply by piperfw** [reply]
That makes sense. Let me know if there is any simple docs/wiki I might be able to contribute to.
